### PR TITLE
[Debug] Support showing exceptions as plain text

### PR DESCRIPTION
**Pull request by c960657**
| Q | A |
| --- | --- |
| Branch? | master |
| Bug fix? | no |
| New feature? | yes |
| BC breaks? | no |
| Deprecations? | yes |
| Tests pass? | yes |
| Fixed tickets | -
| License | MIT |
| Doc PR | - |

The exception listener displays a pretty HTML page. This is nice when developing web pages but less pratical when making a JSON API, working in the terminal etc.

In some cases you can just omit enabling the Symfony exception listener under these circumstances and just use PHP's built-in error messages, but that makes exceptions be handled by a very different code path. The server does not send a `Content-Type: text/plain` header, so the stack trace looks garbled if you view them in a regular browser (outside the view source window). And sometimes you don't register the class with set_exception_handler() but simply calls it to get a nicely formatted version of an exception, e.g. in [Silex\ExceptionHandler->onSilexError](https://github.com/silexphp/Silex/blob/1.3/src/Silex/ExceptionHandler.php#L45).

This patch adds support for generating plain-text output in ExceptionHandler. The output is more similar to the HTML version than that of PHP's built-in error messages.

The patch also adds a some detection logic that tries to guess whether HTML or plain-text is appropriate in a given request. This is just a guess but hopefully a better one than always assuming HTML.

While moving the HTML generation code to HtmlFormatter I made a few minor changes:
- I removed the vendor prefixes for the border-radius CSS rules. These are supported by [all modern browsers](http://caniuse.com/border-radius).
- I changed the htmlspecialchars() flag from ENT_QUOTES (escape both single and double quotes) to ENT_COMPAT (escape double quotes, not single quotes). This avoids some unnecessary escaping (makes it easier to read in view source mode). Escaping single quotes is only required for values enclosed in single quoted HTML attributes - all our attributes are double-quoted.
- I added a `<title>` element, making the output valid HTML5.

### CODE REVIEW

File: src/Symfony/Component/Debug/Formatter/Formatter.php [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\Debug\Formatter;
+
+use Symfony\Component\Debug\Exception\FlattenException;
+
+interface Formatter
```
Comment (phansys) [inline src/Symfony/Component/Debug/Formatter/Formatter.php]: Missing `Interface` suffix.

File: src/Symfony/Component/Debug/Formatter/TextFormatter.php [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+    protected function formatClass($class)
+    {
+        return $class;
+    }
+
+    protected function formatPath($path, $line)
+    {
+        $path = $this->sanitizeString($path);
+
+        return sprintf('in %s:%d', $path, $line);
```
Comment (phansys) [inline src/Symfony/Component/Debug/Formatter/TextFormatter.php]: IMO second placeholder should be replaced with `%u`.

File: src/Symfony/Component/Debug/Debug.php [hunk 3]
```diff
@@ -47,7 +52,19 @@ public static function enable($errorReportingLevel = E_ALL, $displayErrors = tru
 
         if ('cli' !== PHP_SAPI) {
             ini_set('display_errors', 0);
-            ExceptionHandler::register();
+
+            // Default to HTML, unless requested with a text-only browser or with XMLHttpRequest.
+            if ($formatter === null) {
```
Comment (phansys) [inline src/Symfony/Component/Debug/Debug.php]: Please, use Yoda condition.

File: src/Symfony/Component/Debug/Formatter/Formatter.php [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+     * Sets the charset used by exception messages.
+     *
+     * @param string $charset The charset used by exception messages.
+     */
+    public function setCharset($string);
+
+    /**
+     * Gets the MIME type of the content returned by this formatter.
+     *
+     * @param string A MIME-type, possibly with a charset parameter.
```
Comment (phansys) [inline src/Symfony/Component/Debug/Formatter/Formatter.php]: This method hasn't any argument. I think this is the description for the `@return` docblock.

File: src/Symfony/Component/Debug/Formatter/HtmlFormatter.php [hunk 5]
```diff
@@ -0,0 +1,262 @@
...
+                    }
+
+                    $content .= "    </ol>\n</div>\n";
+                }
+            } catch (\Exception $e) {
+                $title = sprintf('Exception thrown when handling an exception (%s: %s)', get_class($e), $this->escapeHtml($e->getMessage()));
+            }
+        }
+
+        return <<<EOF
```
Comment (phansys) [inline src/Symfony/Component/Debug/Formatter/HtmlFormatter.php]: Please, use nowdoc instead.

Comment (c960657) [inline src/Symfony/Component/Debug/Formatter/HtmlFormatter.php]: @phansys Thanks for the review. I have fixed all the issues you have mentioned.

File: src/Symfony/Component/Debug/Formatter/FormatterInterface.php [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+     *
+     * @return string A MIME-type, possibly with a charset parameter.
+     */
+    public function getContentType();
+
+    /**
+     * Gets the formatted exception.
+     *
+     * @param FlattenException The exception to format.
+     * @param bool             Whether to output detailed debug information.
```
Comment (phansys) [inline src/Symfony/Component/Debug/Formatter/FormatterInterface.php]: Did you miss the `@return` docblock?

Comment (c960657) [inline src/Symfony/Component/Debug/Formatter/FormatterInterface.php]: Yes, sorry, fixed :)

File: src/Symfony/Component/Debug/Exception/FlattenException.php [hunk 7]
```diff
@@ -47,7 +48,16 @@ public static function create(\Exception $exception, $statusCode = null, array $
             $statusCode = 500;
         }
 
+        switch ($statusCode) {
+            case 404:
+                $title = 'Sorry, the page you are looking for could not be found.';
```
Comment (phansys) [inline src/Symfony/Component/Debug/Exception/FlattenException.php]: If you're proposing this feature for an API, I think the term "page" should be reworded. I really don't know which could be better: "route", "resource", "path", "endpoint", "URI"?

**Comment by nicolas-grekas** [comment]
If I recall correctly, we already refused a PR of mine that ended up echo non-json output in a json response. Which would also be the case here, isn't it?

**Comment by c960657** [comment]
@nicolas-grekas Which PR are you referring to?

File: src/Symfony/Component/Debug/Exception/FlattenException.php (hunk 7 above)
Comment (c960657) [inline src/Symfony/Component/Debug/Exception/FlattenException.php]: "route", "resource" and "endpoint" are very technical terms that most regular users wont understand.

I prefer "URL" (I agree with [this WHATWG spec](https://url.spec.whatwg.org/#goals) that the terms URI and IRI are confusing and not widely used).

Inspired by Google's 404 page I suggest the following:
"Sorry, the requested URL was not found on this server."

I think this is neutral, to the point, and does not use too much tech lingo.

**Comment by nicolas-grekas** [comment]
I can't find it bu basically, it was about allowing `dump()` to generate its output in a json advertised page.

**Comment by c960657** [comment]
I think you are referring to PR #14478, but there isn't much discussion there. If somebody has arguments against this proposal, please post them here.

**Comment by fabpot** [comment]
ping @nicolas-grekas

File: src/Symfony/Component/Debug/Debug.php [hunk 8]
```diff
@@ -28,10 +32,11 @@ class Debug
...
      * class loader is also registered.
      *
-     * @param int  $errorReportingLevel The level of error reporting you want
-     * @param bool $displayErrors       Whether to display errors (for development) or just log them (for production)
+     * @param int                $errorReportingLevel The level of error reporting you want
+     * @param bool               $displayErrors       Whether to display errors (for development) or just log them (for production)
+     * @param FormatterInterface $formatter           The output formatter.
      */
-    public static function enable($errorReportingLevel = E_ALL, $displayErrors = true)
+    public static function enable($errorReportingLevel = E_ALL, $displayErrors = true, FormatterInterface $formatter = null)
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Debug/Debug.php]: I don't think we need add this argument: this is a helper, let's keep it simple, if people want to inject their formatter, they can do the job by themselves.

File: src/Symfony/Component/Debug/Exception/FlattenException.php [hunk 9]
```diff
@@ -24,6 +24,7 @@ class FlattenException
 {
     private $message;
     private $code;
+    private $title;
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Debug/Exception/FlattenException.php]: having a title is another feature not required by this PR, I'd prefer not adding it here but well, in another PR.

File: src/Symfony/Component/Debug/ExceptionHandler.php [hunk 10]
```diff
@@ -13,16 +13,15 @@
 
 use Symfony\Component\Debug\Exception\FlattenException;
 use Symfony\Component\Debug\Exception\OutOfMemoryException;
+use Symfony\Component\Debug\Formatter\FormatterInterface;
+use Symfony\Component\Debug\Formatter\HtmlFormatter;
 
 /**
- * ExceptionHandler converts an exception to a Response object.
+ * ExceptionHandler catches exceptions and generates debug output.
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Debug/ExceptionHandler.php]: It does not catch exceptions, php does. What about 
`ExceptionHandler pretty prints exceptions to ease debugging.`

**Comment by nicolas-grekas** [comment]
I've got the point :) In fact, I like the idea: when html is not the best dumping format, switching to text.
I'm wondering if an implementation based on VarDumper wouldn't be better to both enhance HTMl rendering and allow easily switching to text (or even CLI colored output if we were to use the handler this way btw).

File: src/Symfony/Component/Debug/Exception/FlattenException.php (hunk 9 above)
Comment (fabpot) [inline src/Symfony/Component/Debug/Exception/FlattenException.php]: I agree and I'm not even sure that this is a good idea. A `FlattenException` is not something you want to "configure", this is just a representation of an existing exception.

**Comment by fabpot** [comment]
@c960657 What's the status of this PR?

**Comment by c960657** [comment]
@fabpot Hmm, I kind of forgot about this. Let's see if we can agree on how to proceed.

@nicolas-grekas (this is a response to an old question) I like the thought about just dumping the stack trace with VarDumper, but that would imply a big change to how the stack traces look. Personally I think the output from VarDumper is rather ugly, so this may be a controversial change. I think it is a good idea, but I would like to get some consensus about this before proceeding.

**Comment by fabpot** [comment]
@nicolas-grekas Any feedback?

**Comment by julienfalque** [comment]
What about supporting JSON as well?

**Comment by nicolas-grekas** [comment]
@c960657 still interested in moving forward? This needs a rebase on 3.4, where the HTML output has changed a lot. Maybe you should remove all that part and focus on text/plain only in this PR (and keep the rest for new PRs.)
Please see my previous comments also.

**Comment by c960657** [comment]
@nicolas-grekas I'll give it a shot :-) I can't promise

**Comment by c960657** [comment]
Sorry about closing the PR — fat-finger error.

@nicolas-grekas I have rerolled the patch.

The title stuff is removed, and `Debug::enable()` no longer takes a `$formatter` argument.

The `HtmlFormatter` has been updated with the new HTML layout. But I'm not sure what you meant with _“Maybe you should remove all that part and focus on text/plain only in this PR”_?

**Comment by nicolas-grekas** [comment]
@c960657 up to finish this one for the end of the week? I already rebased it locally but cannot push on your repos, would you mind allowing it (or rebasing?)
Then, there'll be some details I'd like to change before merging.

**Comment by c960657** [comment]
@nicolas-grekas I have granted you access to my repo. Let me know if you need anything else from me.

**Comment by nicolas-grekas** [comment]
Looking more carefully at the proposed implementation, I think we should not do this move.
I actually spend some time rebasing the code on latest 3.4, then trying to open/close the new interfaces as strictly as I could in order to easy future maintenance. Then I gave up. Not because we cannot do it nor because the current is behind (that's not the case at all, the code is really fine.)

But the reason is that I still have VarDumper in mind. I read the output of VarDumper may not be looking nice to everyone, yet I still think this is a much better move. It should be noted that the output improved a lot in 3.4 for exception traces. The inspection capabilities in VarDumper far superior than what we might achieve here in the future (because we just won't do it twice and the amount of work put in VarDumper is already huge.)

So, instead of adding a new concept in ExceptionHandler (the FormatterInterface), I'd really prefer to work on using VarDumper in Debug, and improving it if needed.

I do share the goal of having a nice non-HTML output. CliDumper provides it already.

@c960657 sorry about this, and thank you for your work!

**Comment by nicolas-grekas** [comment]
Replaced by issue #24354 so we won't forget about it.

**Pull request closed without merging**